### ISSUE TITLE
LTO fails combining modules with and without opaque pointers

### ISSUE THREAD

**Issue by fhahn** [llvm/llvm-project#55377]
It looks like doing LTO doesn't work when some modules use opaque pointers and some don't. I *think* up to now doing LTO with modules generated by older versions of LLVM worked.

a.ll:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
define i8* @foo() {
    ret i8* null
}
```

b.ll:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
define ptr @bar2() {
    ret ptr null
}
```

```
bin/llvm-as a.ll
bin/llvm-as b.ll
bin/llvm-lto2 run a.bc b.bc -o res.bc -r a.bc,foo,plx -r b.bc,bar2,plx                      <<<
llvm-lto2: b.bc: Opaque pointers are only supported in -opaque-pointers mode (Producer: 'LLVM15.0.0git' Reader: 'LLVM 15.0.0git')
```

At the moment, this is causing plenty of `Bitcode` tests to fail from `llvm-test-suite`, when building the test suite with LTO.

**Comment by nikic** [comment]
This probably depends on the order in which the files are linked: It should work if an opaque pointer module comes first.

There's principally two ways I see of fixing this. The first one is to add a pre-scan phase that goes through all bitcode modules and determined whether they use opaque pointers or not (without materializing any IR) and then enabled opaque pointer mode based on that.

The second one is to default LTO to use opaque pointers. Using opaque pointers with typed pointer modules works fine, but the converse is not true.

We should probably go with the second one, given how we'd do that anyway soon enough. I guess that would involve adding some `no-opaque-pointers` options to lld and LLVMgold.so and then making clang pass those if necessary.

**Comment by fhahn** [comment]
>The second one is to default LTO to use opaque pointers. Using opaque pointers with typed pointer modules works fine, but the converse is not true.
>
> We should probably go with the second one, given how we'd do that anyway soon enough. I guess that would involve adding some no-opaque-pointers options to lld and LLVMgold.so and then making clang pass those if necessary.

Yeah that sounds good. Given that this option would only be temporary it might be good enough for most cases to just set an option in the LTO config based on the CMake configuration. 

The option would also be needed in `llvm-lto`, `llvm-lto2`, `libLTO`. Possibly others too.

**Comment by nikic** [comment]
We already have the normal `-opaque-pointers` option that should be usable for llvm-lto and llvm-lto2 at least.

And from a quick look, it looks like at least LLVMgold will just pass through unknown options to ParseCommandLineOptions(). So I think we might not actually need any changes in LTO utilities, and just make the clang driver pass the existing option.

**Comment by fhahn** [comment]
Thanks for the fix!

**Comment by badger200** [comment]
This breaking problem still exists on Darwin using Apple’s ld64-609 linker. I can’t find a method to pass opaque-pointers=1 to ld64 when building clang. If I switch to ld64.lld, it links successfully.

I’m currently building clang 15.0.7 with clang 15.0.3.

**Comment by badger200** [comment]
Finally a solution for clang-15 on Darwin using ld64! Here it is!
I have spent so many hours chasing this obnoxious bug introduced with clang-15. The trick I needed was to add it to clang command line via:
`-Wl,-mllvm,--opaque-pointers`
Finally success!!! This seems to be the only true method to get clang to tell ld64 to use libLTO.dylib with opaque pointers enabled regardless.
